### PR TITLE
[browserperfdash-benchmark] Support supplying browser arguments on linux

### PR DESCRIPTION
**Pull request by spectranaut**
#### 51cfd93f9ff47129799f25676bcca75e48f57416
<pre>
[browserperfdash-benchmark] Support supplying browser arguments on linux
<a href="https://bugs.webkit.org/show_bug.cgi?id=256555">https://bugs.webkit.org/show_bug.cgi?id=256555</a>

Reviewed by Carlos Alberto Lopez Perez and Elliott Williams.

Add support for &quot;--browser-args&quot; on linux. This change involved changing
the browser arg arguments from a positional args to non-positional args.

Canonical link: <a href="https://commits.webkit.org/264122@main">https://commits.webkit.org/264122@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/9b9d8ff2195c010bb54f68c6a99a2d2e2d9040f2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6790 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7009 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7193 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8380 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7043 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6789 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/8182 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6952 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/9939 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6908 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/7507 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6231 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8475 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/6895 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/4934 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6149 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/13956 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/6667 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6230 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/9013 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6719 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/5506 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6103 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1602 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10281 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6480 "Built successfully") | | |

### CODE REVIEW

File: Tools/Scripts/webkitpy/benchmark_runner/run_benchmark.py [hunk 1]
```diff
@@ -60,12 +60,13 @@ def config_argument_parser():
...
     parser.add_argument('--generate-pgo-profiles', dest="generate_pgo_profiles", action='store_true', help="Collect LLVM profiles for PGO, and copy them to the diagnostics directory.")
     parser.add_argument('--profile', action='store_true', help="Collect profiling traces, and copy them to the diagnostic directory.")
-    parser.add_argument('browser_args', nargs='*', help='Additional arguments to pass to the browser process')
 
     group = parser.add_mutually_exclusive_group()
     group.add_argument('--browser-path', help='Specify the path to a non-default copy of the target browser as a path to the .app.')
     group.add_argument('--build-directory', dest='build_dir', help='Path to the browser executable (e.g. WebKitBuild/Release/).')
 
+    parser.add_argument('--browser-args', nargs=argparse.REMAINDER, help='Additional arguments to pass to the browser process. If using, this argument MUST be the last argument in the list (all sequental arguments or strings will be considered browser arguments).')
+
```
Comment (clopez) [inline Tools/Scripts/webkitpy/benchmark_runner/run_benchmark.py]: This changes the CLI interface for the script `run-benchmark`

Is this change really needed? I would prefer if we can avoid breaking current hypothetical users of this script that are passing the browser args without the switch `--browser-args`  (note: I'm speculating, I don't know how everyone uses this script)

File: Tools/Scripts/webkitpy/browserperfdash/browserperfdash_runner.py [hunk 2]
```diff
@@ -243,7 +245,17 @@ def run(self):
...
+                                                    self._args.localCopy,
+                                                    self._args.countOverride,
+                                                    self._args.timeoutOverride,
+                                                    self._args.buildDir,
+                                                    temp_result_file.name,
+                                                    self._args.platform,
+                                                    self._args.browser,
+                                                    None, None, None, None,
+                                                    None, None, None, None,
+                                                    self._args.browser_args)
```
Comment (clopez) [inline Tools/Scripts/webkitpy/browserperfdash/browserperfdash_runner.py]: Instead of passing all those `None` parameters, maybe this can be simplified by simply passing `browser_args=self._args.browser_args` ?

File: Tools/Scripts/webkitpy/benchmark_runner/run_benchmark.py (hunk 1 above)
Comment (spectranaut) [reply Tools/Scripts/webkitpy/benchmark_runner/run_benchmark.py]: right actually I couldn't use the positional arguments -- at least, I couldn't figure out a way for them to work and I did try. The problem is that the browser args start with `--` so argsparse tries to interpret them as  arguments... it doesn't know when to stop trying to interpret things as arguments and instread interpret as values, so you have to use an explicit argument: `--browser-args` and the remaining items will be considered values

@emw-apple added this feature very recently (~month ago) and I meant to already cc him -- Elliot is this change ok? are you using this?

File: Tools/Scripts/webkitpy/browserperfdash/browserperfdash_runner.py (hunk 2 above)
Comment (spectranaut) [reply Tools/Scripts/webkitpy/browserperfdash/browserperfdash_runner.py]: oh! can I do that in python? I'll try

File: Tools/Scripts/webkitpy/benchmark_runner/run_benchmark.py (hunk 1 above)
Comment (clopez) [reply Tools/Scripts/webkitpy/benchmark_runner/run_benchmark.py]: You can pass arguments with dashes after two empty dashes. Try this:
```
Tools/Scripts/run-benchmark --plan sunspider --count 1 --browser chrome -- --arguments --for --chrome
```
Then  `browser_args` should be `['--arguments', '--for', '--chrome']`

This is not obvious so it is maybe a good idea to add some note about how to pass the arguments on the help text

Comment (spectranaut) [reply Tools/Scripts/webkitpy/benchmark_runner/run_benchmark.py]: oh nice -- in this case I can leave the old API and update the description

Comment (spectranaut) [reply Tools/Scripts/webkitpy/benchmark_runner/run_benchmark.py]: actually since this is not an urgent change I think I'll wait for a week or so and see if @emw-apple is ok with the change that adds the explicit `--browser-args`, out of a slight preference for me to have the explicit `--browser-args`. But if that positional argument is being used anywhere I'll change my PR to use the old API, like I said above

Comment (emw-apple) [reply Tools/Scripts/webkitpy/benchmark_runner/run_benchmark.py]: We are using the positional form internally. The reason I wrote it this way was that I felt it was more ergonomic than having to pass multiple arguments nested in a single argument, e.g. `run-benchmark --browser-args="--arguments --for --chrome"` because we'd then have to worry about nested levels of escaping.  If there's a compelling reason to switch to this approach that's okay with me—it'd be easy to adapt our internal usage—but I would prefer to keep it as is :)

> You can pass arguments with dashes after two empty dashes. Try this:
> 
> ```
> Tools/Scripts/run-benchmark --plan sunspider --count 1 --browser chrome -- --arguments --for --chrome
> ```
> 
> Then `browser_args` should be `['--arguments', '--for', '--chrome']`
> 
> This is not obvious so it is maybe a good idea to add some note about how to pass the arguments on the help text

+1, I should have done this earlier!

Comment (spectranaut) [reply Tools/Scripts/webkitpy/benchmark_runner/run_benchmark.py]: reverted to use positional args! thanks for chimming in @emw-apple

**Review by emw-apple — APPROVED** [review]

**Review by clopez — APPROVED** [review]
